File: infra/faxbot/plugin-dev-kit/node/testing.js
const os = require('os');
const fs = require('fs');
const path = require('path');

function createMockDeps() {
  const logger = new Proxy({}, { get: () => () => {} });
  const storage = {}; const db = {}; const http = {}; const audit = {}; const cache = {}; const metrics = {};
  const events = { _events: {}, emit(e, p = {}) { this._events[e] = this._events[e] || []; this._events[e].push(p); }, getEvents(e) { return this._events[e] || []; } };
  const configDir = fs.mkdtempSync(path.join(os.tmpdir(), 'faxbot_plugin_'));
  return { logger, storage, db, events, http, audit, config_dir: configDir, cache, metrics, extras: {} };
}

class PluginTestHarness {
  constructor(PluginClass) { this.PluginClass = PluginClass; this.plugin = null; this.deps = null; }
  async setup(config = null) { this.deps = createMockDeps(); this.plugin = new this.PluginClass(this.deps); if (config) { if (this.plugin.validateConfig) this.plugin.validateConfig(config); if (this.plugin.initialize) await this.plugin.initialize(config); } return this; }
  async teardown() { if (this.plugin && this.plugin.shutdown) await this.plugin.shutdown(); if (this.deps && this.deps.config_dir) { try { fs.rmSync(this.deps.config_dir, { recursive: true, force: true }); } catch {} } }
  assertEventEmitted(event) { const list = this.deps.events.getEvents(event); if (!list || !list.length) throw new Error(`Event ${event} was not emitted`); }
  assertNoPhiLogged() { /* No-op (stub logger) */ }
}

module.exports = { createMockDeps, PluginTestHarness };

File: infra/faxbot/plugin-dev-kit/python/faxbot_plugin_dev/testing.py
from typing import Dict, Any
from unittest.mock import MagicMock
import tempfile
import shutil
import asyncio

from .base import PluginDeps, PluginManifest


def create_mock_deps() -> PluginDeps:
    logger = MagicMock()
    storage = MagicMock()
    db = MagicMock()
    events = MagicMock()
    http = MagicMock()
    audit = MagicMock()
    cache = MagicMock()
    metrics = MagicMock()

    events._events = {}

    def emit(event: str, payload: Dict[str, Any] = None):
        payload = payload or {}
        events._events.setdefault(event, []).append(payload)

    def get_events(event: str):
        return events._events.get(event, [])

    events.emit = emit
    events.get_events = get_events

    config_dir = tempfile.mkdtemp(prefix="faxbot_plugin_")

    return PluginDeps(
        logger=logger,
        storage=storage,
        db=db,
        events=events,
        http=http,
        audit=audit,
        config_dir=config_dir,
        cache=cache,
        metrics=metrics,
        extras={},
    )


class PluginTestCase:
    def __init__(self, plugin_class):
        self.plugin_class = plugin_class
        self.plugin = None
        self.deps = None

    async def setup(self, config: Dict[str, Any] = None):
        self.deps = create_mock_deps()
        self.plugin = self.plugin_class(self.deps)
        if config:
            self.plugin.validate_config(config)
            await maybe_await(self.plugin.initialize(config))
        return self

    async def teardown(self):
        if self.plugin:
            await maybe_await(self.plugin.shutdown())
        if self.deps and getattr(self.deps, 'config_dir', None):
            shutil.rmtree(self.deps.config_dir, ignore_errors=True)

    def assert_event_emitted(self, event: str):
        events = self.deps.events.get_events(event)
        assert events, f"Event {event} was not emitted"

    def assert_no_phi_logged(self):
        for call in self.deps.logger.mock_calls:
            args = str(call)
            import re
            if re.search(r'\d{10,}', args):
                assert "***" in args, "PHI (phone number) detected in logs"


async def maybe_await(result):
    if asyncio.iscoroutine(result):
        return await result
    return result

